lib/src/resource_owner_password_grant.dart

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library resource_owner_password_grant;
+
+import 'dart:async';
+import 'dart:convert';
+import 'package:http/http.dart' as http;
+import 'package:crypto/crypto.dart';
+import 'client.dart';
+import 'handle_access_token_response.dart';
+import 'utils.dart';
+
+/// Implementation of the [resource owner password grant] (http://tools.ietf.org/html/draft-ietf-oauth-v2-31#section-4.3) for oauth 2.
+///
+
+/// Returns a fully-authorized [Client] if authorization is successful.
+///
+/// The client can provide a [clientId] and [clientSecret] for authenticating itself  as required by the server. The
+/// default authentication is basic authentication as recommended by the spec. This can be overridden to be passed as
+/// query parameters by passing [useBasicAuth]: false.
+///
+/// Specific scopes can be requested vis [scopes], but is not required.  The server may choose to grant less scopes than
+/// actually requested.  The actual scopes granted are returned in [Credentials] property of the [Client].
+///
+Future<Client> resourceOwnerPasswordGrant(
+    Uri authorizationEndpoint, String username, String password,
+    {String clientId, String clientSecret: '', List<String> scopes: const [],
+    bool useBasicAuth: true, http.Client httpClient}) async {
```

Comment:
I'd like to keep `useBasicAuth` out of this PR. I don't want to add it here if it's not also in `AuthorizationCodeGrant`, and that's a job for a separate patch.